File: web_interface/validate_xml.py
#!/usr/bin/env python3
"""
XML Validation Tool for Neural VM Builder
Loads an XML file, sends it to the backend for code generation and execution,
then reports any runtime errors found.
"""

import sys
import requests
import json
from pathlib import Path

BACKEND_URL = "http://localhost:5000"

def load_and_generate_code(xml_path):
    """
    Load XML file and use selenium/playwright to generate Python code.
    For now, we'll just instruct the user to use the web interface.
    """
    print(f"Loading XML from: {xml_path}")

    xml_file = Path(xml_path)
    if not xml_file.exists():
        print(f"❌ Error: XML file not found: {xml_path}")
        return None

    print("\n⚠️  This tool requires the web interface to be open.")
    print("Please:")
    print("1. Open web_interface/index.html in your browser")
    print("2. Load the XML file using the '📁 Load' button")
    print("3. Click the '▶️ Run Code' button")
    print("4. Check the execution output panel for errors")

    return None

def validate_xml_via_api(xml_path):
    """Send XML path to backend for validation"""
    try:
        response = requests.post(
            f"{BACKEND_URL}/validate-xml",
            json={"xml_path": str(xml_path)},
            timeout=10
        )
        return response.json()
    except requests.exceptions.ConnectionError:
        print(f"❌ Error: Cannot connect to backend at {BACKEND_URL}")
        print("Make sure the backend is running: python backend.py")
        return None
    except Exception as e:
        print(f"❌ Error: {e}")
        return None

def main():
    if len(sys.argv) < 2:
        print("Usage: python validate_xml.py <xml_file_path>")
        print("Example: python validate_xml.py simple_gpt_example.xml")
        sys.exit(1)

    xml_path = sys.argv[1]

    print("=" * 60)
    print("Neural VM Builder - XML Validation Tool")
    print("=" * 60)

    # Try API approach first
    result = validate_xml_via_api(xml_path)

    if result and result.get('success'):
        print("✅ Validation passed!")
        print(f"\nOutput:\n{result.get('stdout', '')}")
    elif result:
        print(f"❌ Validation failed: {result.get('error', 'Unknown error')}")
        if result.get('stderr'):
            print(f"\nError details:\n{result.get('stderr')}")
    else:
        # Fall back to manual instructions
        load_and_generate_code(xml_path)

if __name__ == "__main__":
    main()

File: web_interface/validate_xml.js
#!/usr/bin/env node
/**
 * XML Validation Tool for Neural VM Builder
 * Loads an XML file, generates Python code using Blockly, and executes it
 * to catch runtime errors.
 */

const fs = require('fs');
const { execSync } = require('child_process');
const path = require('path');

// Check if XML file is provided
if (process.argv.length < 3) {
    console.log('Usage: node validate_xml.js <xml_file_path>');
    console.log('Example: node validate_xml.js simple_gpt_example.xml');
    process.exit(1);
}

const xmlPath = process.argv[2];

console.log('='.repeat(60));
console.log('Neural VM Builder - XML Validation Tool');
console.log('='.repeat(60));

// Check if file exists
if (!fs.existsSync(xmlPath)) {
    console.log(`❌ Error: XML file not found: ${xmlPath}`);
    process.exit(1);
}

console.log(`📁 Loading XML: ${xmlPath}`);

// Read XML content
const xmlContent = fs.readFileSync(xmlPath, 'utf8');

console.log('🔄 Generating Python code from XML...\n');

// Generate code by sending POST request to backend
const http = require('http');

const postData = JSON.stringify({ xml_content: xmlContent });

const options = {
    hostname: 'localhost',
    port: 5000,
    path: '/validate-xml-content',
    method: 'POST',
    headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData)
    }
};

const req = http.request(options, (res) => {
    let data = '';

    res.on('data', (chunk) => {
        data += chunk;
    });

    res.on('end', () => {
        try {
            const result = JSON.parse(data);

            if (result.success) {
                console.log('✅ Validation passed!');
                console.log(`\n📊 Execution time: ${result.execution_time}s`);
                if (result.stdout) {
                    console.log(`\n📤 Output:\n${result.stdout}`);
                }
            } else {
                console.log(`❌ Validation failed!`);
                console.log(`\n⚠️  Error: ${result.error}`);
                if (result.stderr) {
                    console.log(`\n📋 Details:\n${result.stderr}`);
                }
            }
        } catch (e) {
            console.log(`❌ Failed to parse response: ${e.message}`);
        }
    });
});

req.on('error', (e) => {
    console.log(`❌ Error: Cannot connect to backend at http://localhost:5000`);
    console.log('Make sure the backend is running: python backend.py');
    process.exit(1);
});

req.write(postData);
req.end();